vllm/entrypoints/openai/api_server.py

```diff
@@ -619,6 +620,28 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
     else:
         served_model = args.model
 
+    if args.chat_template is not None:
+        try:
+            with open(args.chat_template, "r") as f:
+                content = f.read()
+                try:
+                    json_data = json.loads(content)
+                    if "chat_template" in json_data:
+                        chat_template = json_data["chat_template"]
+                    else:
+                        chat_template = content
+                except json.JSONDecodeError:
+                    # If JSON fails, use the file content as raw text
+                    chat_template = content
+        except OSError as e:
+            try:
+                # If opening a file fails, set chat template to be args to
+                # ensure we decode so our escape are interpreted correctly
+                chat_template = codecs.decode(args.chat_template,
+                                              "unicode_escape")
```

Comment:
shouldn't we use jinja as well? seems like HF uses jinja template right?

Reply:
This is still jinja, it just properly parses the escaped characters (`\n` and the like).

It allows users to pass in the template like so:
```
python3 -m vllm.entrypoints.openai.api_server --model Tostino/Inkbot-13B-8k-0.2 --tensor-parallel-size 2 --chat-template "<#meta#>\n- Date: {{ (messages|selectattr('role', 'equalto', 'meta-current_date')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'meta-current_date')|list) else '' }}\n- Task: {{ (messages|selectattr('role', 'equalto', 'meta-task_name')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'meta-task_name')|list) else '' }}\n<#system#>\n{{ (messages|selectattr('role', 'equalto', 'system')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'system')|list) else '' }}\n<#chat#>\n{% for message in messages %}\n{% if message['role'] == 'user' %}\n<#user#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% elif message['role'] == 'assistant' %}\n<#bot#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% elif message['role'] == 'user_context' %}\n<#user_context#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% endif %}\n{% endfor %}\n{% if add_generation_prompt and messages[-1]['role'] != 'assistant' %}\n<#bot#>\n{% endif %}"
```

Reply:
For chat template I think it is good to recommend for users to write it into a file, rather than maintaining on the CLI. No need to do two things when there is one good way of doing it 😄

Reply:
Are you sure we should lock in with file-on-disk with all the more recent "serverless" type deployment methods, container images, etc?  It allows the model's template to be copy/pasted from the tokenizer_config.json into the CLI argument, and modified as necessary without any extra steps. It's easy enough to remove, just seemed like a nice QOL improvement.

Reply:
You are right. Actually dumping it into a envvar wouldn't be a bad idea overall.

Reply:
Agreed there. It alright to do that in another PR? There are likely other envvars we would want to add support for at the same time.

Reply:
I think this is pretty inline with this feature right? cc @WoosukKwon for input in this here.

Reply:
Is this conversation alright to mark "resolved"?

Reply:
I think file is a good first pass here.